juju/juju
# Introduce and use global clock in lease manager

## Description of change

This PR introduces a global clock document to the database. Each controller machine agent runs a worker that will periodically (1s) attempt to advance the global time by the same (period) amount. Concurrent updates will be prevented, and unsuccessful workers will back off for a short (30s) delay. We guarantee that the global time is monotonically increasing, and increases at a rate no faster than wall clock time.

We pass a global clock reader into the state/lease client, so that it can compute expiry times based on the global time. Lease documents store the start time as a global time, and the request duration; we then make the expiry time relative to the local time for use by clients. The local time is expected to contain a monotonic component (i.e. Go 1.9+), so it can be compared to time.Now().

We refresh the global clock time when a claim or extension is made, to guarantee that the lease is held for at least the specified duration. When a lease is to be expired, we first attempt to use the most recent global clock time, and then refresh if the lease expires after the cached time, only failing if the lease expires after the refreshed time.

## QA steps

1. juju bootstrap localhost
2. juju enable-ha
3. juju deploy ubuntu -n 3

Identify the lease holders for "singular-controller" and "ubuntu", and stop those agents (controller machine agent and unit agent respectively). Another agent should eventually claim the leases.

Also:

1. juju bootstrap localhost (with juju 2.2.5)
2. juju deploy postgresql -n 3
(wait for postgresql units to become idle, take note of the leader)
3. juju enable-ha
(wait for additional controllers to be voting)
4. juju upgrade-juju -m controller (with this branch)

Verify that:
 - there are no old docs in the leases collection
 - the leases are kept by the previous holders across upgrade

## Documentation changes

None.

## Bug reference

Fixes https://bugs.launchpad.net/juju/+bug/1706340
